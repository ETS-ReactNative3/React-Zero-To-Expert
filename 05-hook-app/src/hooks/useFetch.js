// Version 3
import { useState, useEffect, useRef } from 'react'

const useFetch = (url) => {
  const isMounted = useRef(true)
  const [state, setState] = useState({ data: null, loading: true, error: null })

  useEffect(() => {
    return () => {
      // Se va disparar cuando el componente o efecto se desmonte
      // Aqui cambiamos el valor del useRef
      // Esto no va disparar nuevamente la renderizacion del componente
      isMounted.current = false
    }
  }, [])

  useEffect(() => {
    setState({ data: null, loading: true, error: null })
    fetch(url)
      .then((resp) => resp.json())
      .then((data) => {
        if (isMounted.current) {
          setState({
            loading: false,
            error: null,
            data,
          })
        }
      })
  }, [url])

  return state
}

export default useFetch

// // Version 2
// import { useState, useEffect, useRef } from 'react'

// const useFetch = (url) => {
//   const isMounted = useRef(true)
//   const [state, setState] = useState({ data: null, loading: true, error: null })

//   useEffect(() => {
//     return () => {
//       // Se va disparar cuando el componente o efecto se desmonte
//       // Aqui cambiamos el valor del useRef
//       // Esto no va disparar nuevamente la renderizacion del componente
//       isMounted.current = false
//     }
//   }, [])

//   useEffect(() => {
//     setState({ data: null, loading: true, error: null })
//     fetch(url)
//       .then((resp) => resp.json())
//       .then((data) => {
//         setTimeout(() => {
//           if (isMounted.current) {
//             setState({
//               loading: false,
//               error: null,
//               data,
//             })
//           } else {
//             console.log('setState no se llamo')
//           }
//         }, 4000)
//       })
//   }, [url])

//   return state
// }

// export default useFetch

console.log('#####')

// // Version 1
// import { useState, useEffect } from 'react'

// const useFetch = (url) => {
//   const [state, setState] = useState({ data: null, loading: true, error: null })

//   useEffect(() => {
//     setState({ data: null, loading: true, error: null })
//     fetch(url)
//       .then((resp) => resp.json())
//       .then((data) => {
//         setTimeout(() => {
//           setState({
//             loading: false,
//             error: null,
//             data,
//           })
//         }, 4000)
//         // setState({
//         //   loading: false,
//         //   error: null,
//         //   data,
//         // })
//       })
//   }, [url])

//   return state
// }

// export default useFetch
