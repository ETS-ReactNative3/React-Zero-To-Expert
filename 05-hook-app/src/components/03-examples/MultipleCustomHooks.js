import React from 'react'
import useFetch from '../../hooks/useFetch'
import useCounter from './useCounter'

const MultipleCustomHooks = () => {
  const { counter, increment } = useCounter(1)
  // const state = useFetch(`https://www.breakingbadapi.com/api/quotes/6`)
  // console.log(state)
  // const { loading, data } = useFetch(
  //   `https://www.breakingbadapi.com/api/quotes/12 `
  // )
  const { loading, data } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  )
  console.log(loading)
  console.log(data)
  // null, !null negacion de null es true, !!null doble negacion de null es false
  const { author, quote } = !!data && data[0]
  console.log(author, quote)
  return (
    <div>
      <h1>MultipleCustomHooks</h1>
      <h2>BreakingBad Quotes</h2>
      <hr />
      {loading ? (
        <div className='alert alert-info text-center'>Loading...</div>
      ) : (
        <blockquote className='blockquote text-end'>
          <p className='mb-3'>{quote}</p>
          <footer className='blockquote-footer'>{author}</footer>
        </blockquote>
      )}
      <button className='btn btn-primary' onClick={increment}>
        Siguiente Quote
      </button>
    </div>
  )
}

export default MultipleCustomHooks
