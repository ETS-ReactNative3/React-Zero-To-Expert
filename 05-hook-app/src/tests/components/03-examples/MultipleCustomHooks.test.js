import React from 'react'
import { shallow } from 'enzyme'
import MultipleCustomHooks from '../../../components/03-examples/MultipleCustomHooks'
import useFetch from '../../../hooks/useFetch'
// import useCounter from '../../../components/03-examples/useCounter'
import useCounter from '../../../components/03-examples/useCounter'

jest.mock('../../../hooks/useFetch')
jest.mock('../../components/03-examples/useCounter')

describe('Pruebas en <MultipleCustomHooks />', () => {
  useCounter.mockReturnValue({
    counter: 10,
    increment: () => {},
  })

  test('debe de mostrarse correctamente', () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null,
    })

    const wrapper = shallow(<MultipleCustomHooks />)
    expect(wrapper).toMatchSnapshot()
  })
  test('debe de mostrar la informacion', () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: ' ElfGod',
          quote: 'Elf Leader',
        },
      ],
      loading: false,
      error: null,
    })
    const wrapper = shallow(<MultipleCustomHooks />)
    expect(wrapper.find('.alert').exists()).toBe(false)
    expect(wrapper.find('.mb-3').text().trim()).toBe('Elf Leader')
    expect(wrapper.find('footer').text().trim()).toBe('ElfGod')
    console.log(wrapper.html())
  })
})
