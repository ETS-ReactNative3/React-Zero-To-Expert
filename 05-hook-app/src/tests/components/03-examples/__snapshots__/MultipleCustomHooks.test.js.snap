// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks /> debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    MultipleCustomHooks
  </h1>
  <h2>
    BreakingBad Quotes
  </h2>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente Quote
  </button>
</div>
`;
